app/websocket/websocketclient.c

```diff
@@ -0,0 +1,816 @@
+// Websocket client
+
+#include "osapi.h"
+#include "user_interface.h"
+#include "espconn.h"
+#include "mem.h"
+#include "limits.h"
+#include "stdlib.h"
+
+#include "c_types.h"
+#include "c_string.h"
+#include "c_stdlib.h"
+#include "c_stdio.h"
+
+#include "websocketclient.h"
+
+// Depends on 'crypto' module for sha1
+#include "../crypto/digests.h"
+#include "../crypto/mech.h"
+
+#define PROTOCOL_SECURE "wss://"
+#define PROTOCOL_INSECURE "ws://"
+
+#define PORT_SECURE 443
+#define PORT_INSECURE 80
+#define PORT_MAX_VALUE 65535
+
+#define SSL_BUFFER_SIZE 5120
+
+// TODO: user agent configurable
+#define WS_INIT_HEADERS  "GET %s HTTP/1.1\r\n"\
+                         "Host: %s:%d\r\n"\
+                         "Upgrade: websocket\r\n"\
+                         "Connection: Upgrade\r\n"\
+                         "User-Agent: ESP8266\r\n"\
+                         "Sec-Websocket-Key: %s\r\n"\
+                         "Sec-WebSocket-Protocol: chat\r\n"\
+                         "Sec-WebSocket-Version: 13\r\n"\
+                         "\r\n"
+
+#define WS_INIT_HEADERS_LENGTH 169
+#define WS_GUID "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+#define WS_GUID_LENGTH 36
+
+#define WS_HTTP_SWITCH_PROTOCOL_HEADER "HTTP/1.1 101"
+#define WS_HTTP_SEC_WEBSOCKET_ACCEPT "Sec-WebSocket-Accept:"
+
+#define WS_CONNECT_TIMEOUT_MS 10 * 1000
+#define WS_PING_INTERVAL_MS 30 * 1000
+#define WS_FORCE_CLOSE_TIMEOUT_MS 5 * 1000
+#define WS_UNHEALTHY_THRESHOLD 2
+
+#define WS_OPCODE_CONTINUATION 0x0
+#define WS_OPCODE_TEXT 0x1
+#define WS_OPCODE_BINARY 0x2
+#define WS_OPCODE_CLOSE 0x8
+#define WS_OPCODE_PING 0x9
+#define WS_OPCODE_PONG 0xA
+
+static char *cryptoSha1(char *data, unsigned int len) {
+  SHA1_CTX ctx;
+  SHA1Init(&ctx);
+  SHA1Update(&ctx, data, len);
+  
+  uint8_t *digest = (uint8_t *) os_zalloc(20);
+  SHA1Final(digest, &ctx);
+  return (char *) digest; // Requires free
+}
+
+static const char *bytes64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
```

Comment:
Interesting. I supposed it could make some sense...

But at this moment this `websocketclient` isn't coupled to Lua, and I think it's better that way.

I thought about exposing some interface (`ifdefs` maybe, or a static global variable) that allowed the `malloc` to be overridden. That way, I could change it to use the luaM_malloc (which can potentially enforce a gc in desperate situations). Maybe this could also be fine tuned as well...